realm/realm-library/src/androidTest/java/io/realm/RealmObjectTests.java

```diff
@@ -1668,4 +1670,107 @@ public void removeChangeListeners_throwOnUnmanagedObject() {
             looperThread.testComplete();
         }
     }
+
+    // Bug https://github.com/realm/realm-java/issues/2569
+    @Test
+    @RunTestInLooperThread
+    public void addChangeListener_returnedObjectOfCopyToRealmOrUpdate() {
+        looperThread.realm.beginTransaction();
+        looperThread.realm.createObject(AllTypesPrimaryKey.class, 1);
+
+        AllTypesPrimaryKey allTypesPrimaryKey = new AllTypesPrimaryKey();
+        allTypesPrimaryKey.setColumnLong(1);
+        allTypesPrimaryKey.setColumnFloat(42f);
+        allTypesPrimaryKey = looperThread.realm.copyToRealmOrUpdate(allTypesPrimaryKey);
+        looperThread.realm.commitTransaction();
+
+        allTypesPrimaryKey.addChangeListener(new RealmChangeListener<AllTypesPrimaryKey>() {
+            @Override
+            public void onChange(AllTypesPrimaryKey element) {
+                assertEquals(42.0f, element.getColumnFloat(), 0f);
+                looperThread.testComplete();
+            }
+        });
+    }
+
+    // The object should be added to HandlerController.realmObjects only when the first time addListener called.
+    @Test
+    @RunTestInLooperThread
+    public void addChangeListener_shouldAddTheObjectToHandlerRealmObjects() {
+        looperThread.realm.beginTransaction();
+        AllTypesPrimaryKey allTypesPrimaryKey = looperThread.realm.createObject(AllTypesPrimaryKey.class, 1);
+        looperThread.realm.commitTransaction();
+        assertTrue(looperThread.realm.handlerController.realmObjects.isEmpty());
+
+        allTypesPrimaryKey.addChangeListener(new RealmChangeListener<RealmModel>() {
```

Comment:
please use `AllTypesPrimaryKey` instead of `RealmModel`. We can't infer the type correctly for `realmObject.addChangeListener` this is why we use `<T extends RealmModel>` in the  method definition